go,goBuildModules: bump default go version to 1.18

SuperSandro2000 opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 committed:
go,goBuildModules: bump default go version to 1.18

SuperSandro2000 committed:
cod: mark broken

SuperSandro2000 commented:
broken on target branch:
- cod
- python310Packages.ipfshttpclient

SuperSandro2000 committed:
cosign: pin to go 1.17

SuperSandro2000 committed:
mole: pin to go 1.17

SuperSandro2000 committed:
garble: pin to go 1.17

SuperSandro2000 committed:
dendrite: pin to go 1.17

SuperSandro2000 committed:
hyprspace: pin to go 1.17

SuperSandro2000 committed:
gomplate: pin to go 1.17

SuperSandro2000 committed:
gophernotes: pin to go 1.17

SuperSandro2000 committed:
livedl: pin to go 1.17

SuperSandro2000 committed:
go-chromecast: pin to go 1.17

SuperSandro2000 committed:
go-mockery: pin to go 1.17

SuperSandro2000 committed:
grafana-agent: pin to go 1.17

SuperSandro2000 committed:
vegeta: pin to go 1.17

SuperSandro2000 committed:
gowitness: pin to go 1.17

SuperSandro2000 committed:
kconf: pint to go 1.17

SuperSandro2000 committed:
marathonctl: pin to go 1.17

SuperSandro2000 committed:
prometheus-nats-exporter: pin to go 1.17

rski commented:
This fixes addlicense, https://github.com/google/addlicense/commit/8cfc1b5425e5884193fa88f3d96dce680ea2514c, I'll open a separate PR to upgrade it to an unstable version as it hasn't released in a year+

rski commented:
as promised: https://github.com/NixOS/nixpkgs/pull/179666

rski commented:
go-langserver is "deprioritised" in favour of gopls. Maybe it should just be removed in nixpkgs: https://github.com/sourcegraph/go-langserver
```
Note: We have deprioritized work on this language server for use in editors in favor of Google's Go language server, [gopls](https://github.com/golang/tools/tree/master/gopls). It is in the best interests of the community to only have a single language server.
```

rski commented:
https://github.com/sjas/gosca gosca is also deprecated, archived, also possible candidate for deletion. I'm mostly suggesting deleting packages that won't be made to work upstream with 1.18 because 1.19 is around the corner, at which point 1.17 will no longer be supported. Keeping packages that can't be upgraded (or don't make sense to be upgraded) to work with 1.18 at that point will not be that good.

rski commented:
ipget: https://github.com/NixOS/nixpkgs/pull/179670

Will get to the others at some other point

rski commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1007,7 +1007,9 @@ with pkgs;
   addic7ed-cli = callPackage ../tools/misc/addic7ed-cli { };
 
-  addlicense = callPackage ../tools/misc/addlicense { };
+  addlicense = callPackage ../tools/misc/addlicense {
```
fixed by https://github.com/NixOS/nixpkgs/pull/179666

rski commented on pkgs/top-level/all-packages.nix:
```diff
@@ -23838,7 +23838,9 @@ with pkgs;
   gomodifytags = callPackage ../development/tools/gomodifytags { };
 
-  go-langserver = callPackage ../development/tools/go-langserver { };
+  go-langserver = callPackage ../development/tools/go-langserver {
```
go-langserver is "deprioritised" in favour of gopls. Given that it will likely never work with go1.18 and generics, maybe it should just be removed in nixpkgs, see https://github.com/sourcegraph/go-langserver

rski commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2557,7 +2557,9 @@ with pkgs;
   gopass-summon-provider = callPackage ../tools/security/gopass/summon.nix { };
 
-  gosca = callPackage ../development/tools/gosca { };
+  gosca = callPackage ../development/tools/gosca {
```
https://github.com/sjas/gosca gosca is also deprecated, archived, also possible candidate for deletion. I'm mostly suggesting deleting packages that won't be made to work upstream with 1.18 because 1.19 is around the corner, at which point 1.17 will no longer be supported. Keeping packages that can't be upgraded (or don't make sense to be upgraded) to work with 1.18 at that point will not be that good.

rski commented on pkgs/top-level/all-packages.nix:
```diff
@@ -7376,7 +7376,9 @@ with pkgs;
   ipfs-migrator = callPackage ../applications/networking/ipfs-migrator { };
 
-  ipget = callPackage ../applications/networking/ipget { };
+  ipget = callPackage ../applications/networking/ipget {
```
https://github.com/NixOS/nixpkgs/pull/179670

rski commented:
(moved my comments inline, duh)

SuperSandro2000 commented:
https://github.com/NixOS/nixpkgs/pull/179709

SuperSandro2000 committed:
unconvert: pin to go 1.17

SuperSandro2000 committed:
toxiproxy: pin to go 1.17

rski commented:
This looks fine to me, I'll let someone else give the final approval. I do think we can either fix or drop all the remaining packages, that can be left as future work over the next couple of weeks before 1.19 comes out

zowoq commented:
Have you checked all `vendorSha256` hashes are reproducible with 1.18? 

What are you planning to do about all darwin breakage on x86_64? Leaving contributors to figure out how to deal with dozens of broken and unfixable packages doesn't seem like a good idea. I'd suggest doing a hydra run just for `x86_64-darwin` and marking them as broken as part of this PR.

SuperSandro2000 commented:
> Have you checked all `vendorSha256` hashes are reproducible with 1.18?

I checked some and then determined that they won't change after none of my test examples changed.

> What are you planning to do about all darwin breakage on x86_64?

I pinged them in the darwin channel 5 days ago, 3 days ago and just now. So far I didn't receive any feedback. The Go 1.18 update is already stalled for a good amount of time and I don't want to make it another project held back by proprietary darwin. Lets wait a few days if someone provides feedback, otherwise I would just move ahead.

SuperSandro2000 committed:
photon: fix hash

SuperSandro2000 committed:
proto-contrib: fix hash, remove redundand platforms

SuperSandro2000 committed:
sqlc: fix vendor hash, fix meta.platforms

SuperSandro2000 committed:
pigeon: fix vendor hash

SuperSandro2000 committed:
vgo2nix: remove myself from maintainers

SuperSandro2000 committed:
vgo2nix: mark broken

SuperSandro2000 committed:
protoc-gen-twirp_typescript: fix vendor hash

SuperSandro2000 committed:
protolock: fix vendor hash

SuperSandro2000 committed:
endlessh-go: fix vendor hash

SuperSandro2000 committed:
telegraf: fix vendor hash

SuperSandro2000 committed:
zsh-history: fix vendor hash, formatting

SuperSandro2000 committed:
opentelemetry-collector: fix vendor hash

SuperSandro2000 commented:
I fixed all vendor hashes. I didn't double check every one but I assume some of them where already broken in master but they are in total not enough to be a generic issue of the 1.18 update.

I also missed a message in the darwin matix channel. Someone is probably proposing two PRs in the next days.

dasJ commented:
I didn't see this PR and already started to script the task :/ The script is currently running and is generating new vendor hashes when needed.
While I believe my approach with scripting this task is more future-proof in the long run, I actually prefer this PR because it fixes additional things as well.
Once the script is finished, I will open a (draft) PR as well so you can check if you missed any hashes. I'm still investigating if I can automatically mark packages as broken on Darwin but I will keep you posted.

toonn committed:
go_1_18: Use apple_sdk_11_0.callPackage

Go 1.18 requires a newer SDK than the one we build from sources. As a
workaround we're making use of the SDK we're using for aarch64-darwin.
This means Go 1.18 will not work on any Darwin systems that don't have
forwards-compatible SDK versions with the particular package in
question. We might need to mark Go packages broken based on the macOS
version rather than just the platform and architecture.

Until we find a better solution, Go packages will need to make sure to
get all their (Darwin) system dependencies from the `apple_sdk_11_0`,
this includes dependencies of build tools like `xcbuild`.

For convenience `darwin.apple_sdk_11_0` has a `callPackage` attribute
which provides the correct `stdenv` and `xcbuild` attributes as
arguments. This function can be expanded to substitute other necessary
arguments when they come up.

SuperSandro2000 commented:
Between mine and your comment dasJ is a PR linked which hopefully fixes most of the darwin build problems.

I am planning to rebase this PR at least once more, so I would prefer to cherry pick extra fixes into here or send them directly to master if they already apply there.

> While I believe my approach with scripting this task is more future-proof in the long run, I actually prefer this PR because it fixes additional things as well.

I encountered some problems for example with envoy which I didn't know how to solve and the packages where (more) broken after the hash update, so I ignored them.

dasJ commented:
I'm in luck, then. I started my own branch because I didn't see you PR which mean I already have the Darwin change in my worktree \o/

dasJ commented:
Alright, everything broke because my store overflowed… This is the script I used, maybe it's helpful for future updates:
```bash
#!/usr/bin/env bash

# Requirements for this to work:
# - Run this on x86_64-linux
# - Have at least one x86_64-darwin builder configured
# - Run this script in your nixpkgs checkout

# Do this before running the script:
# - Bump go, buildGoPackage, buildGoModule in all-packages.nix
# - Remove the overrides to the new version in all-packages.nix (for example `buildGoModule = buildGo1XXModule`)

# Do this after running the script:
# - Commit the changes
# - Somehow handle the packages that were reported as broken by this script
# - Run nixpkgs-review or something for packages that use `go` rather than `buildGoModule` or `buildGoPackage`

set -euo pipefail
export NIX_PATH= # better safe than sorry

echo -e '\e[31mPhase 1: Finding packages that use go modules...\e[0m'
nix-env --option system x86_64-linux -f "${PWD}" -qaP --json --show-trace > pkgs.json
goPkgsWithModules="$(nix eval --raw --impure --expr 'let pkgs = import ./. {}; in pkgs.lib.concatStringsSep "\n" (builtins.filter (x: pkgs.lib.isAttrs (pkgs.lib.attrByPath ((pkgs.lib.splitString "." x) ++ [ "go-modules" ]) "" pkgs)) (builtins.attrNames (builtins.fromJSON (builtins.readFile ./pkgs.json))))')"
echo "Found $(wc -l <<< "${goPkgsWithModules}" || :) packages"

echo -e '\e[31mPhase 2: Fixing module hashes...\e[0m'
broken=()
for package in ${goPkgsWithModules}; do
	echo "Trying to fetch modules of ${package}..."
	if ! out="$(nix-build --expr "(import ./. {}).${package}.go-modules.overrideAttrs (_:{name=\"test\";})" 2>&1)"; then
		if ! grep -q 'hash mismatch in fixed-output derivation' <<< "${out}"; then
			echo "Package seems broken"
			broken+=("${package}")
			continue
		fi
		if ! file="$(nix eval --raw --impure --expr "(builtins.unsafeGetAttrPos \"${package}\" (import ./. {})).file" 2>&1)"; then
			file="$(find pkgs -type f -path "*${package}/default.nix" | head -n1)"
		fi
		if [[ "${file}" = *'pkgs/top-level/all-packages.nix' ]]; then
			file="$(find pkgs -type f -path "*${package}/default.nix" | head -n1)"
		fi
		if ! [[ -f "${file}" ]]; then
			echo "Could not find source file"
			broken+=("${package}")
			continue
		fi

		echo "Fixing module hash in ${file}..."
		correctHash="$(grep got: <<< "${out}" | tail -n1 | awk '{print $2}' | xargs nix hash to-sri)"
		sed -i "s:vendorSha256 = \"[^\"]*\":vendorSha256 = \"${correctHash}\":g" "${file}"
		if git diff --exit-code "${file}" &>/dev/null; then
			echo "Hash was not properly fixed"
			broken+=("${package}")
			continue
		fi
		git add "${file}"
	fi
done
echo "Packages that have broken vendor hashes and cannot be fixed: ${broken[*]}"

echo -e '\e[31mPhase 3: Finding all go packages without modules...\e[0m'
goPkgsWithoutModules="$(nix eval --raw --impure --expr 'let pkgs = import ./. {}; in pkgs.lib.concatStringsSep "\n" (builtins.filter (x: pkgs.lib.hasAttrByPath ((pkgs.lib.splitString "." x) ++ [ "goPackagePath" ]) pkgs) (builtins.attrNames (builtins.fromJSON (builtins.readFile ./pkgs.json))))')"
echo "Found $(wc -l <<< "${goPkgsWithoutModules}" || :) packages"

echo -e '\e[31mPhase 4: Building on x86_64-linux...\e[0m'
broken2=()
for package in ${goPkgsWithModules} ${goPkgsWithoutModules}; do
	# shellcheck disable=SC2076
	if [[ " ${broken[*]} " =~ " ${package} " ]]; then
		echo "Skipping ${package} as it's already broken"
		continue
	fi

	echo "Trying to build ${package}..."
	if ! nix-build -A "${package}" &>/dev/null; then
		echo "Failed"
		broken2+=("${package}")
	fi
done
echo "Packages that do not build on x86_64-linux: ${broken2[*]}"

echo -e '\e[31mPhase 5: Finding Darwin packages...\e[0m'
nix-env --option system x86_64-darwin -f "${PWD}" -qaP --json --show-trace > pkgs.json
darwinPkgs="$(nix eval --raw --impure --expr 'let pkgs = import ./. {}; in pkgs.lib.concatStringsSep "\n" (builtins.filter (x: pkgs.lib.isAttrs (pkgs.lib.attrByPath ((pkgs.lib.splitString "." x) ++ [ "go-modules" ]) "" pkgs)) (builtins.attrNames (builtins.fromJSON (builtins.readFile ./darwin-pkgs.json))))')"
echo "Found $(wc -l <<< "${darwinPkgs}" || :) packages"

echo -e '\e[31mPhase 6: Trying to build on x86_64-darwin...\e[0m'
broken3=()
for package in ${goPkgsWithModules} ${goPkgsWithoutModules}; do
	# shellcheck disable=SC2076
	if [[ " ${broken[*]} " =~ " ${package} " ]]; then
		echo "Skipping ${package} as the module hash is not fixable"
		continue
	fi
	# shellcheck disable=SC2076
	if [[ " ${broken2[*]} " =~ " ${package} " ]]; then
		echo "Skipping ${package} as it's already broken on Linux"
		continue
	fi
	# shellcheck disable=SC2076
	if ! [[ " ${darwinPkgs[*]} " =~ " ${package} " ]]; then
		echo "Skipping ${package} as it's not supported on Darwin"
		continue
	fi

	if ! nix-build -A "${package}" &>/dev/null; then
		echo "Failed"
		broken3+=("${package}")
	fi
done
echo "Packages that do not build on x86_64-darwin: ${broken3[*]}"
```

SuperSandro2000 commented:
> pkgs/applications/networking/cluster/argocd-autopilot/default.nix | 2 +-
>  pkgs/development/libraries/protolock/default.nix                  | 2 +-
>  pkgs/development/tools/database/sqlc/default.nix                  | 2 +-
>  pkgs/development/tools/pigeon/default.nix                         | 2 +-
>  pkgs/development/tools/proto-contrib/default.nix                  | 2 +-
>  pkgs/development/tools/protoc-gen-twirp_typescript/default.nix    | 2 +-
>  pkgs/development/tools/vgo2nix/default.nix                        | 2 +-
>  pkgs/development/tools/wails/default.nix                          | 2 +-
>  pkgs/servers/endlessh-go/default.nix                              | 2 +-
>  pkgs/servers/http/pomerium/default.nix                            | 2 +-
>  pkgs/servers/monitoring/telegraf/default.nix                      | 2 +-
>  pkgs/shells/hilbish/default.nix                                   | 2 +-
>  pkgs/shells/zsh/zsh-history/default.nix                           | 2 +-
>  pkgs/tools/misc/steampipe/default.nix                             | 2 +-

Double checked those and didn't find any new vendor hash mismatches.

zowoq commented:
> I am planning to merge this mid next week.

`go` still needs to be changed to point to `go_1_18` as part of this PR so `go` and `buildGo{Module,Package}` all use the same version.

These packages with broken modules on 1.18 still need to be fixed or pinned:

> Packages where the modules don't build with reasons that are not related to broken hashes: `alertmanager-bot deepsea exercism filtron git-hound gobetween httpdump kcli kfctl kubectl-doctor kubectl-evict-pod kubemqctl magnetico ncdns opensnitch opentelemetry-collector-contrib operator-sdk pam_ussh pg_featureserv photon-rss prometheus-nginxlog-exporter safe shadowfox shhgit ssb` ... `waitron wrap wuzz zfsbackup zrepl`

___

The `terraform-providers` changes in c3892557e9eb24d8aee3b4ddb33f333b26fcda9a can be dropped from this PR, I removed them as part of 77b2ff803d29a1d86b08333bf228fc751aad477e.

SuperSandro2000 commented:
> `go` still needs to be changed to point to `go_1_18`

done

> These packages with broken modules on 1.18 still need to be fixed or pinned:

Those failed due to a full disk:

> Alright, everything broke because my store overflowed…

zowoq commented:
> Those failed due to a full disk ...

No. I have checked several of them and they are actually broken.

SuperSandro2000 commented:
> No. I have checked several of them and they are actually broken.

The vendoring falls apart when doing ``--rebuild`` because upstreams go.mod needs changes. Pinning those packages to 1.17 now.

SuperSandro2000 committed:
shadowfox: pin to go 1.17, update vendor hash, complete ldflags, shorten description

SuperSandro2000 committed:
waitron: pin to go 1.17

SuperSandro2000 committed:
git-hound: pin to go 1.17

SuperSandro2000 committed:
httpdump: pin to go 1.17

SuperSandro2000 committed:
kfctl: pin to go 1.17

SuperSandro2000 committed:
opensnitch: pin to go 1.17

SuperSandro2000 committed:
safe: pin to go 1.17

SuperSandro2000 committed:
ssb: pin to go 1.17

SuperSandro2000 committed:
wrap: pin to go 1.17

SuperSandro2000 committed:
wuzz: pin to go 1.17

SuperSandro2000 committed:
zfsbackup: pin to go 1.17

SuperSandro2000 committed:
filtron: pin to go 1.17

SuperSandro2000 committed:
sshgit: pin to go 1.17

SuperSandro2000 committed:
gobetween: pin to go 1.17

SuperSandro2000 committed:
ncdns: mark broken

SuperSandro2000 committed:
prometheus-nginxlog-exporter: pin to go 1.17

SuperSandro2000 committed:
pg_featureserv: pin to go 1.17

SuperSandro2000 committed:
alertmanager-bot: pin to go 1.17

SuperSandro2000 committed:
pam_ussh: pin to go 1.17

SuperSandro2000 committed:
exercism: pin to go 1.17

SuperSandro2000 committed:
kubectl-evict-pod: pin to go 1.17

SuperSandro2000 committed:
kubemqctl: pin to go 1.17

SuperSandro2000 committed:
kubectl-doctor: pin to go 1.17

SuperSandro2000 committed:
magnetico: pin to go 1.17

SuperSandro2000 committed:
deepsea: pin to go 1.17

SuperSandro2000 committed:
zrepl: pin to go 1.17

SuperSandro2000 committed:
kcli: pin to go 1.17

SuperSandro2000 commented:
So, this should be good to go now. If we find more broken packages we can fix them after a merge.

chasecaleb commented:
> So, this should be good to go now. If we find more broken packages we can fix them after a merge.

@SuperSandro2000 Trivy is one of them and the breakage is causing me problems at work. Any chance of adding that in? See:
- https://github.com/NixOS/nixpkgs/pull/173793
- https://github.com/NixOS/nixpkgs/blob/8c6b75e3b87a91098f27158893022adaa648d299/pkgs/tools/admin/trivy/default.nix#L52-L54

willcohen commented:
`podman` is another broken-on-darwin-x86 that hopefully unbreaks after this!

06kellyjac commented:
> Trivy is one of them and the breakage is causing me problems at work

If it's not removed in this PR I'll make a PR straight after (excluding when I'm sleeping) :slightly_smiling_face:

chasecaleb commented:
> > Trivy is one of them and the breakage is causing me problems at work
>
> Edit: in fact reading through this thread I might be able to drop it now because of #176661

Yeah, that's what I'm thinking too. I haven't tried myself though. Do you need any help with this? I have lots of other things going on but if needed I can make time. Feel free to ping me in a PR or whatever is convenient for you.

zowoq commented:
> Would a generic comment be good enough?
> 
> ```
> # pinned due to build failure or vendoring problems. When unpinning double check with: nix-build -A $name.go-modules --rebuild
> ```

Yes.

SuperSandro2000 commented:
> If it's not removed in this PR I'll make a PR straight

Yes, please. I am not touching already marked as broken packages in this PR otherwise it would grow completely out of hand.

SuperSandro2000 committed:
treewide: add a comment explaining to buildGo117Module override

SuperSandro2000 commented:
Here goes nothin!

rski commented:
the nix-rebuild comment doesn't look right to me, the --rebuild flag seems to only be part of `nix build`

SuperSandro2000 commented:
really? ufff...
